lean4.html.markdown

```diff
@@ -0,0 +1,587 @@
+  let p4 : 2^8 * (2 - 1) = 2^8 :=
+    Nat.mul_one (2^8)
+  Eq.trans (Eq.trans p2 p3) p4
+
+/-
+For individual steps, we use proofs already available in Lean. For example:
+
+  `Nat.pow_succ a b` is the proof of `a^(b + 1) = a^b * a`.
+
+We use some general mechanisms in the above proof that are trivial in written,
+English mathematics.
+
+`congrArg f p`is a proof of `f a = f b` from `a = b`.
+
+  Notice how we use `p1` to build `p2`. Indeed, `p2` is just the rewrite proved
+  by `p1` in a complicated function.
+
+`Eq.symm p` takes a proof `p` of `a=b` and gives us a proof of `b=a`.
+
+The left distributivity is defined as `a * (b - c) = a * b - a * c`. The
+equality we want to prove is its symmetric version.
+-/
+
+/-
+We don't want to write proofs in such detail. We would like to specify them at a
+high-level. This is what we do when we write mathematics in English.
+
+Enter *tactics*. These are commands to Lean that guide it to construct proofs by
+itself.
+
+A common tactic is `simp`, short for simplify.
+
+We can ask to Lean to try to prove by simplifying using `by simp`. The `by`
+tells Lean that what follows is a tactic, and not a complete proof.
+
+Then, the proof becomes:
+-/
+def two_power_eight' : 2^9 - 2^8 = 2^8 := by
+  simp
+
+/-
+We can use `theorem` instead of `def` as a notational convenience.
+-/
+theorem two_power_eight : 2^9 - 2^8 = 2^8 := by simp
+
+/-
+The `calc` primitive allows us to prove equalities using stepwise
+calculations. Each step has to be justified by a proof.
+-/
+theorem plus_squared (a b : Nat) : (a+b)^2 = a^2 + 2*a*b + b^2 :=
+  calc
+    (a+b)^2 = (a+b)*(a+b)             := Nat.pow_two (a+b)
+    _       = (a+b)*a + (a+b)*b       := Nat.mul_add (a+b) a b
+    _       = a*a + b*a + (a*b + b*b) := by repeat rw [Nat.add_mul]
+    _       = a*a + b*a + a*b + b*b   := by rw [← Nat.add_assoc]
+    _       = a*a + a*b + a*b + b*b   := by rw [Nat.mul_comm b a]
+    _       = a^2 + a*b + a*b + b*b   := by rw [← Nat.pow_two a]
+    _       = a^2 + a*b + a*b + b^2   := by rw [← Nat.pow_two b]
+    _       = a^2 + (a*b + a*b) + b^2 := by rw [Nat.add_assoc (a^2)]
+    _       = a^2 + 2*(a*b) + b^2     := by rw [← Nat.two_mul (a*b)]
+    _       = a^2 + 2*a*b + b^2       := by rw [Nat.mul_assoc 2 a b]
+
+/-
+Let us now prove more "realistic" theorems. Those involving logical connectives.
+
+First, we define even and odd numbers.
+-/
+def Even (n : Nat) := ∃ k, n = 2*k
+def Odd  (n : Nat) := ∃ k, n = 2*k + 1
+
+/-
+To prove an existential, we can provide specific values if we know them.
+-/
+theorem zero_even : Even 0 :=
+  let h : 0 = 2 * 0 := Eq.symm (Nat.mul_zero 2)
+  Exists.intro 0 h
+/-
+`Exists.intro v h` proves `∃ x, p x` by substituting `x` by `v` and using the
+proof `h` for `p v`.
+-/
+
+/-
+Now, we will see how to use hypothesis that are existentials to prove
+conclusions that are existentials.
+-/
+theorem even_mul_even_is_even' (hn : Even n) (hm : Even m) : Even (n*m) :=
```

Comment:
One should probably bind the `m` and `n` and not rely on Lean's autoimplicit in this expository piece.
Something like `theorem even_mul_even_is_even' {m n} ...`, or with types, or with an explicit `∀ {m n}`.

Reply:
This is probably all around the doc.